File: claude-code/extractors_go/pkg/extract/markdown.go
// Package extract provides HTML to markdown conversion.
package extract

import (
	"fmt"
	"regexp"
	"strings"

	"golang.org/x/net/html"
)

// HtmlToMarkdown converts an HTML string to markdown format.
func HtmlToMarkdown(htmlContent string) string {
	doc, err := html.Parse(strings.NewReader(htmlContent))
	if err != nil {
		return htmlContent
	}

	var b strings.Builder
	convertNode(&b, doc, 0)

	result := b.String()
	result = cleanMarkdown(result)
	return strings.TrimSpace(result)
}

func convertNode(b *strings.Builder, n *html.Node, depth int) {
	switch n.Type {
	case html.TextNode:
		text := n.Data
		if strings.TrimSpace(text) != "" {
			b.WriteString(text)
		}
	case html.ElementNode:
		convertElement(b, n, depth)
	case html.DocumentNode:
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			convertNode(b, c, depth)
		}
	}
}

func convertElement(b *strings.Builder, n *html.Node, depth int) {
	tag := strings.ToLower(n.Data)

	// Skip non-content elements
	switch tag {
	case "script", "style", "noscript", "iframe", "svg", "nav", "header", "footer":
		return
	}

	switch tag {
	case "h1":
		b.WriteString("\n\n# ")
		writeChildren(b, n, depth)
		b.WriteString("\n\n")
	case "h2":
		b.WriteString("\n\n## ")
		writeChildren(b, n, depth)
		b.WriteString("\n\n")
	case "h3":
		b.WriteString("\n\n### ")
		writeChildren(b, n, depth)
		b.WriteString("\n\n")
	case "h4":
		b.WriteString("\n\n#### ")
		writeChildren(b, n, depth)
		b.WriteString("\n\n")
	case "h5":
		b.WriteString("\n\n##### ")
		writeChildren(b, n, depth)
		b.WriteString("\n\n")
	case "h6":
		b.WriteString("\n\n###### ")
		writeChildren(b, n, depth)
		b.WriteString("\n\n")
	case "p":
		b.WriteString("\n\n")
		writeChildren(b, n, depth)
		b.WriteString("\n\n")
	case "br":
		b.WriteString("\n")
	case "hr":
		b.WriteString("\n\n---\n\n")
	case "strong", "b":
		b.WriteString("**")
		writeChildren(b, n, depth)
		b.WriteString("**")
	case "em", "i":
		b.WriteString("*")
		writeChildren(b, n, depth)
		b.WriteString("*")
	case "code":
		if parentTag(n) != "pre" {
			b.WriteString("`")
			writeChildren(b, n, depth)
			b.WriteString("`")
		} else {
			writeChildren(b, n, depth)
		}
	case "pre":
		lang := getCodeLanguage(n)
		b.WriteString(fmt.Sprintf("\n\n```%s\n", lang))
		writeChildren(b, n, depth)
		b.WriteString("\n```\n\n")
	case "a":
		href := getAttr(n, "href")
		b.WriteString("[")
		writeChildren(b, n, depth)
		b.WriteString(fmt.Sprintf("](%s)", href))
	case "img":
		alt := getAttr(n, "alt")
		src := getAttr(n, "src")
		b.WriteString(fmt.Sprintf("![%s](%s)", alt, src))
	case "ul", "ol":
		b.WriteString("\n")
		writeListItems(b, n, depth, tag == "ol")
		b.WriteString("\n")
	case "li":
		// Handled by writeListItems
		writeChildren(b, n, depth)
	case "blockquote":
		b.WriteString("\n\n")
		var inner strings.Builder
		writeChildren(&inner, n, depth)
		for _, line := range strings.Split(inner.String(), "\n") {
			b.WriteString("> " + line + "\n")
		}
		b.WriteString("\n")
	case "table":
		b.WriteString("\n\n")
		writeTable(b, n)
		b.WriteString("\n\n")
	default:
		writeChildren(b, n, depth)
	}
}

func writeChildren(b *strings.Builder, n *html.Node, depth int) {
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		convertNode(b, c, depth+1)
	}
}

func writeListItems(b *strings.Builder, n *html.Node, depth int, ordered bool) {
	idx := 1
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		if c.Type == html.ElementNode && strings.ToLower(c.Data) == "li" {
			indent := strings.Repeat("  ", depth)
			if ordered {
				b.WriteString(fmt.Sprintf("%s%d. ", indent, idx))
				idx++
			} else {
				b.WriteString(fmt.Sprintf("%s- ", indent))
			}
			writeChildren(b, c, depth+1)
			b.WriteString("\n")
		}
	}
}

func writeTable(b *strings.Builder, n *html.Node) {
	var rows [][]string

	var walkTable func(*html.Node)
	walkTable = func(node *html.Node) {
		if node.Type == html.ElementNode && strings.ToLower(node.Data) == "tr" {
			var cells []string
			for c := node.FirstChild; c != nil; c = c.NextSibling {
				if c.Type == html.ElementNode && (strings.ToLower(c.Data) == "td" || strings.ToLower(c.Data) == "th") {
					var cellBuf strings.Builder
					writeChildren(&cellBuf, c, 0)
					text := strings.TrimSpace(cellBuf.String())
					text = strings.ReplaceAll(text, "|", "\\|")
					cells = append(cells, text)
				}
			}
			if len(cells) > 0 {
				rows = append(rows, cells)
			}
			return
		}
		for c := node.FirstChild; c != nil; c = c.NextSibling {
			walkTable(c)
		}
	}
	walkTable(n)

	if len(rows) == 0 {
		return
	}

	maxCols := 0
	for _, row := range rows {
		if len(row) > maxCols {
			maxCols = len(row)
		}
	}

	for i := range rows {
		for len(rows[i]) < maxCols {
			rows[i] = append(rows[i], "")
		}
	}

	b.WriteString("| " + strings.Join(rows[0], " | ") + " |\n")
	separators := make([]string, maxCols)
	for i := range separators {
		separators[i] = "---"
	}
	b.WriteString("| " + strings.Join(separators, " | ") + " |\n")

	for _, row := range rows[1:] {
		b.WriteString("| " + strings.Join(row, " | ") + " |\n")
	}
}

func getAttr(n *html.Node, key string) string {
	for _, a := range n.Attr {
		if strings.ToLower(a.Key) == key {
			return a.Val
		}
	}
	return ""
}

func getCodeLanguage(preNode *html.Node) string {
	// Check for <code> child with language class
	for c := preNode.FirstChild; c != nil; c = c.NextSibling {
		if c.Type == html.ElementNode && strings.ToLower(c.Data) == "code" {
			cls := getAttr(c, "class")
			for _, part := range strings.Fields(cls) {
				for _, prefix := range []string{"language-", "lang-", "highlight-"} {
					if strings.HasPrefix(part, prefix) {
						return strings.TrimPrefix(part, prefix)
					}
				}
			}
			lang := getAttr(c, "data-lang")
			if lang == "" {
				lang = getAttr(c, "data-language")
			}
			if lang != "" {
				return lang
			}
		}
	}
	return ""
}

func parentTag(n *html.Node) string {
	if n.Parent != nil && n.Parent.Type == html.ElementNode {
		return strings.ToLower(n.Parent.Data)
	}
	return ""
}

func cleanMarkdown(md string) string {
	// Remove excessive blank lines
	re := regexp.MustCompile(`\n{4,}`)
	md = re.ReplaceAllString(md, "\n\n\n")

	// Trim trailing whitespace from lines
	lines := strings.Split(md, "\n")
	for i, line := range lines {
		lines[i] = strings.TrimRight(line, " \t")
	}
	return strings.Join(lines, "\n")
}

File: claude-code/extractors_rust/src/markdown.rs
//! HTML to Markdown conversion using the scraper crate.

use scraper::{Html, Node};

/// Convert HTML content to Markdown format.
pub fn html_to_markdown(html: &str) -> String {
    let document = Html::parse_document(html);
    let mut output = String::new();

    convert_node(&document, document.root_element().id(), &mut output);

    clean_markdown(&output)
}

fn convert_node(doc: &Html, node_id: ego_tree::NodeId, output: &mut String) {
    let node = doc.tree.get(node_id).unwrap();

    match node.value() {
        Node::Text(text) => {
            let t = text.trim();
            if !t.is_empty() {
                output.push_str(t);
            }
        }
        Node::Element(el) => {
            let tag = el.name().to_lowercase();

            // Skip non-content elements
            match tag.as_str() {
                "script" | "style" | "noscript" | "iframe" | "svg" | "nav" | "header"
                | "footer" => return,
                _ => {}
            }

            let (prefix, suffix) = get_tag_wrapper(&tag, el);

            output.push_str(&prefix);

            for child in node.children() {
                convert_node(doc, child.id(), output);
            }

            output.push_str(&suffix);
        }
        Node::Document => {
            for child in node.children() {
                convert_node(doc, child.id(), output);
            }
        }
        _ => {}
    }
}

fn get_tag_wrapper(tag: &str, el: &scraper::node::Element) -> (String, String) {
    match tag {
        "h1" => ("\n\n# ".to_string(), "\n\n".to_string()),
        "h2" => ("\n\n## ".to_string(), "\n\n".to_string()),
        "h3" => ("\n\n### ".to_string(), "\n\n".to_string()),
        "h4" => ("\n\n#### ".to_string(), "\n\n".to_string()),
        "h5" => ("\n\n##### ".to_string(), "\n\n".to_string()),
        "h6" => ("\n\n###### ".to_string(), "\n\n".to_string()),
        "p" => ("\n\n".to_string(), "\n\n".to_string()),
        "br" => ("\n".to_string(), String::new()),
        "hr" => ("\n\n---\n\n".to_string(), String::new()),
        "strong" | "b" => ("**".to_string(), "**".to_string()),
        "em" | "i" => ("*".to_string(), "*".to_string()),
        "code" => {
            // Check if parent is <pre>
            // For simplicity, use backtick
            ("`".to_string(), "`".to_string())
        }
        "pre" => {
            let lang = get_code_language(el);
            (format!("\n\n```{}\n", lang), "\n```\n\n".to_string())
        }
        "a" => {
            let href = el.attr("href").unwrap_or("");
            ("[".to_string(), format!("]({})", href))
        }
        "img" => {
            let alt = el.attr("alt").unwrap_or("");
            let src = el.attr("src").unwrap_or("");
            (format!("![{}]({})", alt, src), String::new())
        }
        "li" => ("\n- ".to_string(), String::new()),
        "blockquote" => ("\n\n> ".to_string(), "\n\n".to_string()),
        "ul" | "ol" => ("\n".to_string(), "\n".to_string()),
        _ => (String::new(), String::new()),
    }
}

fn get_code_language(el: &scraper::node::Element) -> String {
    // Check class attribute for language hints
    if let Some(class) = el.attr("class") {
        for cls in class.split_whitespace() {
            for prefix in &["language-", "lang-", "highlight-"] {
                if let Some(lang) = cls.strip_prefix(prefix) {
                    return lang.to_string();
                }
            }
        }
    }

    if let Some(lang) = el.attr("data-lang") {
        return lang.to_string();
    }
    if let Some(lang) = el.attr("data-language") {
        return lang.to_string();
    }

    String::new()
}

fn clean_markdown(md: &str) -> String {
    let mut result = md.to_string();

    // Remove excessive blank lines
    while result.contains("\n\n\n\n") {
        result = result.replace("\n\n\n\n", "\n\n\n");
    }

    // Trim trailing whitespace from each line
    result = result
        .lines()
        .map(|line| line.trim_end())
        .collect::<Vec<_>>()
        .join("\n");

    result.trim().to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_conversion() {
        let html = "<h1>Title</h1><p>Paragraph text here.</p>";
        let md = html_to_markdown(html);
        assert!(md.contains("# Title"));
        assert!(md.contains("Paragraph text here."));
    }

    #[test]
    fn test_code_block() {
        let html = r#"<pre><code class="language-python">def hello():
    pass</code></pre>"#;
        let md = html_to_markdown(html);
        assert!(md.contains("```python"));
        assert!(md.contains("def hello()"));
        assert!(md.contains("```"));
    }

    #[test]
    fn test_links() {
        let html = r#"<a href="https://example.com">Example</a>"#;
        let md = html_to_markdown(html);
        assert!(md.contains("[Example](https://example.com)"));
    }

    #[test]
    fn test_formatting() {
        let html = "<p><strong>bold</strong> and <em>italic</em></p>";
        let md = html_to_markdown(html);
        assert!(md.contains("**bold**"));
        assert!(md.contains("*italic*"));
    }

    #[test]
    fn test_strips_nav() {
        let html = "<nav>Navigation</nav><main><p>Content</p></main>";
        let md = html_to_markdown(html);
        assert!(!md.contains("Navigation"));
        assert!(md.contains("Content"));
    }
}
